_specifications/lsp/3.18/language/completion.md

```diff
@@ -386,6 +399,75 @@ export interface CompletionList {
 		data?: LSPAny;
 	}
 
+	/**
+	 * Specifies how fields from a completion item should be combined with those
+	 * from `completionList.itemDefaults`.
+	 * 
+	 * In unspecified, all fields will be treated as "replace".
+	 * 
+	 * If a field's value is "replace", the value from a completion item will
+	 * always be used instead of the value from `completionItem.itemDefaults`.
+	 * 
+	 * If a field's value is "merge", the values will be merged using the rules
+	 * defined against each field below.
+	 *
+	 * Servers are only allowed to return `applyKind` if the client
+	 * signals support for this via the `completionList.applyKinds`
+	 * capability.
+	 *
+	 * @since 3.18.0
+	 */
+	applyKind?: {
+		/**
+		 * Specifies whether commitCharacters on a completion will replace or be
+		 * merged with those in `completionList.itemDefaults.commitCharacters`.
+		 * 
+		 * If "replace", the commit characters from the completion item will
+		 * always be used unless not provided, in which case those from
+		 * `completionList.itemDefaults.commitCharacters` will be used. An empty
+		 * list can be used if a completion item does not have any commit
+		 * characters and also should not use those from
+		 * `completionList.itemDefaults.commitCharacters`.
+		 * 
+		 * If "merge" the commitCharacters for the completion will be the union
+		 * of all values in both `completionList.itemDefaults.commitCharacters`
+		 * and the completion's own `commitCharacters`. An empty list indicates
+		 * the completion adds no additional commit characters and a value of
+		 * `null` indicates the defaults should not be used.
+		 *
+		 * @since 3.18.0
+		 */
+		commitCharacters?: "replace" | "merge";
+
+		/**
+		 * Specifies whether data on a completion will replace or
+		 * be merged with data from `completionList.itemDefaults.data`.
+		 * 
+		 * If "replace", the data from the completion item will be used if
+		 * provided, otherwise  `completionList.itemDefaults.data` will be used.
+		 * An empty object can be used if a completion item does not have any
+		 * data but also should not use the value from
+		 * `completionList.itemDefaults.data`.
+		 * 
+		 * If "merge", a shallow merge will be performed between
+		 * `completionList.itemDefaults.data` and the completion's own data
+		 * using the following rules:
+		 * 
+		 * - If a completion's data is not specified, the data from
+		 *   `completionList.itemDefaults.data` will be used.
+		 * - If a completion's data is null, the resulting data field will be
+		 *   null (defaults are not used).
```

Comment:
The aim here was so that there was a way to prevent using the default. So if 99.9% of your completion items have some same base data and you want to use `itemDefaults`, you can still opt-out of that default for the 0.1%.

However, in the case of `data`, maybe it's reasonable to just use an empty object (actually I think I wrote that somewhere).

Now you mention it, Dart also doesn't distinguish between `null`/`undefined` in many cases so it might also be problematic there. Lemme take another look over this.